rust/agama-lib/src/base_http_client.rs

```diff
@@ -0,0 +1,133 @@
+use anyhow::Context;
+use reqwest::{header, Client, Response};
+use serde::{de::DeserializeOwned, Serialize};
+
+use crate::{auth::AuthToken, error::ServiceError};
+
+/// Base that all http clients should use.
+///
+/// It provides several features including automatic base url switching,
+/// websocket events listening or object constructions.
+///
+/// Usage should be just thin layer in domain specific client.
+///
+/// ```no_run
+///   use agama_lib::questions::model::Question;
+///   use agama_lib::base_http_client::BaseHTTPClient;
+///   use agama_lib::error::ServiceError;
+///
+///   async fn get_questions() -> Result<Vec<Question>, ServiceError> {
+///     let client = BaseHTTPClient::new()?;
+///     client.get("/questions").await
+///   }
+/// ```
+pub struct BaseHTTPClient {
+    client: Client,
+    pub base_url: String,
+}
+
+const API_URL: &str = "http://localhost/api";
+
+impl BaseHTTPClient {
+    // if there is need for client without authorization, create new constructor for it
+    pub fn new() -> Result<Self, ServiceError> {
+        let token = AuthToken::find().context("You are not logged in")?;
```

Comment:
The error message is correct (I hope) but even better would be a hint how to fix the situation: `"... Try: agama auth login"`?